### PR TITLE
add override RFC

### PR DESCRIPTION
**Pull request by Hunam6**
*briefly discussed [here](https://canary.discord.com/channels/592103645835821068/592115457029308427/989628476011188284)*

### CODE REVIEW

File: rfcs/000-override.md [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+
+# Prior art
+
+As said before, this is used in Java ([doc](https://www.geeksforgeeks.org/the-override-annotation-in-java/)), in Dart ([doc](https://dart.dev/guides/language/type-system#use-sound-parameter-types-when-overriding-methods)), in Kotlin ([doc](https://www.geeksforgeeks.org/overriding-rules-in-kotlin/)) and probably others.
+
+This is also widely used and necessary in Minecraft modding, to override default behaviors of the game.
+
+# Unresolved questions
+
+\-
```
Comment (dumblob) [inline rfcs/000-override.md]: Maybe this important question shall be answered.
```
- Should this be allowed also for built-in methods/structs/functions/...?
```
(I for myself would say yes, it shall allow overwriting of built-ins)

**Comment by ken0x0a** [comment]
I didn't understand why `override` is necessary...

We can do something like following already with [Embedded Struct](https://github.com/vlang/v/blob/master/doc/docs.md#embedded-structs).

```java
public class ExplosiveChickenEntity extends ChickenEntity {
    public ExplosiveChickenEntity(EntityType<? extends ChickenEntity> entityType, World world) {
        super(entityType, world);
    }
    @Override
    public void onDeath(DamageSource damageSource) {
        super.onDeath(damageSource);
        world.createExplosion(this, this.getX(), this.getY(), this.getZ(), 3.5F, Explosion.DestructionType.BREAK);
    }
}
```

is almost same with

```v
struct ChickenEntity {
  world World
}
fn (e &ChickenEntity) onDeath(ds &DamageSource) {}

struct ExplosiveChickenEntity {
  ChickenEntity
}
fn (e &ExplosiveChickenEntity) onDeath(ds &DamageSource) {
  e.ChickenEntity.onDeath(ds)
  e.world.createExplosion(/* ... */)
}
```

**Comment by JalonSolov** [comment]
`@Override` is necessary in Java to make sure the local method gets called.  It's not necessary in V.

File: rfcs/000-override.md [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+
+struct ExplosiveChickenEntity {
+    minecraft_stuff.ChickenEntity
+}
+
+[override: chicken]
+fn (explosive_chicken ExplosiveChickenEntity) on_death(damage_source minecraft_stuff.DamageSource) {
+    chicken.on_death(damage_source)
+    minecraft_stuff.create_explosion(...)
+}
```
Comment (ken0x0a) [inline rfcs/000-override.md:64]: The following already works without `[override]`

```v
fn (explosive_chicken ExplosiveChickenEntity) on_death(damage_source minecraft_stuff.DamageSource) {
    explosive_chicken.ChickenEntity.on_death(damage_source)
    minecraft_stuff.create_explosion(...)
}
```

**Comment by ken0x0a** [comment]
Hmm, I don't understand...

So, what this RFC make V possible?
We can override method without `[override]` already.

**Comment by Hunam6** [comment]
I wasn't aware V allowed to override a method like this. It seems off to me that V allows doing it implicitly, because of readability, how would know just by reading this code that you are overriding a method?
```v
import minecraft

struct ExplosiveChickenEntity {
  minecraft.ChickenEntity
}
fn (e &ExplosiveChickenEntity) onDeath(ds &DamageSource) {
  e.ChickenEntity.onDeath(ds)
  e.world.createExplosion(/* ... */)
}
```

**Comment by ken0x0a** [comment]
V doesn't have syntax for explicitly implements `interface` also,
so I didn't feel unnatural about this.

I originally thought explicit is good for interface,
but V's compiler shows error at compile time and show suggestions about what we need to implement to satisfy the interface and now I feel that's enough...


> how would know just by reading this code that you are overriding a method?

Do we actually need to know whether a method is `override` or not?
Rust and TypeScript also doesn't have any explicit `override` annotation(?).
(Rust actually doesn't have `super()` for `trait impl` as I know.... but has ways to overcome that)

**Comment by dumblob** [comment]
> Do we actually need to know whether a method is `override` or not? Rust and TypeScript also doesn't have any explicit `override` annotation(?). (Rust actually doesn't have `super()` for `trait impl` as I know.... but has ways to overcome that)

If I understood @Hunam6 correctly, then the problem is the **implicit** "shadowing" of imported stuff which actually does **not** require one to write a module name in front of it to shadow/overwrite the method.

Btw. is it really overwrite or just shadow? If it is an overwrite (i.e. "once and forever"), then any module imported later (i.e. after this module we are editing) would not get the original behavior (instead of just the rest of the currently edited module getting the "overwritten" behavior). Wtf? That would be really bad!

What is the expected "scope of validity" of such an overwrite/shadowing? I would expect **only** the ~file~ module in which the overwrite happened (and only **syntactically after** the place where the overwrite happened).

**Comment by JalonSolov** [comment]
V's way of doing things is closer to "implicit shadowing"... except it really isn't.

ExplodingChickenEntity.on_death is different from ChickenEntity.on_death. Although they may have the same signature, they are in different namespaces, not just different contexts.

**Pull request merged**